File: spec/services/friendship_service_spec.rb
require 'rails_helper'

RSpec.describe FriendshipService, type: :service do

  describe "#create_friendship" do
    let!(:user) { create(:user) }
    let!(:friend) { create(:user) }
    let(:service) { FriendshipService }

    subject { service.create_friendship(user, friend) }

    describe "when saving one friendship fails" do
      before { allow_any_instance_of(Friendship).to receive(:save!).and_raise ActiveRecord::Rollback}

      it "returns false" do
        expect(subject).to be false
      end

      it "does not save any friendships" do
        pending "Don't have a way to check this because save! is stubbed"
        fail
      end
    end

    describe "when saving both forward and backward friendships succeeds" do
      it "returns true" do
        expect(subject).to be true
      end

      it "saves two friendships" do
        subject
        expect(Friendship.count).to eq 2
      end
    end
  end
end

File: app/javascript/hello_angular/_services/friendship.service.spec.ts
import { FriendshipService } from './friendship.service';

import { Friendship } from '../_models/friendship';
import { User } from '../_models/user';

describe("FriendshipService", () => {
  let service: FriendshipService;
  let datastore = {
    createRecord: jest.fn()
  };

  it("should be defined", () => {
    service = new FriendshipService(datastore as any);
    expect(service).toBeDefined();
  })

  describe("addFriendship()", () => {
   describe("when adding friend succeeds", () => {
      const friendId = 23;
      let dummyFriend = new User(null, {id: friendId});
      let dummyFriendship = {
        save: jest.fn()
      }
      datastore.createRecord
        .mockReturnValueOnce(dummyFriend)
        .mockReturnValueOnce(dummyFriendship)

      it("calls the Friendship's save function", () => {
        service = new FriendshipService(datastore as any);
        service.addFriendship(friendId);
        expect(dummyFriendship.save).toHaveBeenCalled();
      });
    });
  });
});
